### PR TITLE
change to memoized _uri_canonical fixes #121

### LINKED ISSUE #121: Random t/http-config.t failures
I observe the following kind of failures on some of my smokers:
```
#   Failed test at t/http-config.t line 59.
#          got: 'u:p|slash|.com|GET|not secure|always'
#     expected: '.com|GET|secure|always'

#   Failed test at t/http-config.t line 69.
#          got: 'u:p|slash|.com|success|GET|not secure|always'
#     expected: '.com|success|GET|secure|always'

#   Failed test at t/http-config.t line 73.
#          got: 'u:p|slash|success|GET|not secure|always'
#     expected: 'success|GET|always'
# Looks like you failed 3 tests of 28.
t/http-config.t ............. 
Dubious, test returned 3 (wstat 768, 0x300)
Failed 3/28 subtests 
```
This seems to happen also on other systems, for example: http://www.cpantesters.org/cpan/report/34a5f0ee-13ca-11e9-8c82-aba07c4588f0

### PR DESCRIPTION
**Pull request by doriantaylor**
does what it says on the tin

### CODE REVIEW

File: lib/HTTP/Request.pm [hunk 1]
```diff
@@ -92,7 +92,19 @@ sub uri
...
+
+    my $uri = $self->{_uri};
+
+    if (defined (my $canon = $self->{_uri_canonical})) {
+        # early bailout if these are the exact same string; try to use
+        # the cheapest comparison method possible
+        return $canon if $$canon eq $$uri;
+
+        # otherwise needs refreshing
+        undef $self->{_uri_canonical};
```
Comment (Grinnz) [inline lib/HTTP/Request.pm]: Since it's being set unconditionally in the next statement, this line isn't really needed.

Comment (doriantaylor) [reply lib/HTTP/Request.pm]: ah yes you're right. update forthcoming.

Comment (doriantaylor) [reply lib/HTTP/Request.pm]: and done

**Comment by oalders** [comment]
@vanHoesel does this look good to you?

**Comment by doriantaylor** [comment]
Not my call; [URI #57](https://github.com/libwww-perl/URI/issues/57) that caused this issue was itself a response to some fairly zealous optimization. It is conceivable that these routines actually cost something once upon a time. Indeed, the scheme-specific subclasses are supposed to implement their own `->canonical`, which one might suppose could do anything to return a result.

I have no position on whether the caching in `HTTP::Request` stays or goes; I am only interested in the behaviour of `$uri->canonical`.

**Comment by oalders** [comment]
Thanks @doriantaylor, @Grinnz, @vanHoesel and @skaji!

**Comment by oalders** [comment]
> I would delete the cache.

Let's keep this in mind for the future.  Not sure why it's being cached.

**Comment by oalders** [comment]
Just uploaded a `-TRIAL` release.

**Comment by skaji** [comment]
@oalders Thanks!

**Pull request merged**